components/fxa-client/src/pairing_channel/client_states.rs

```diff
@@ -0,0 +1,388 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use bytes::{Buf, BufMut, BytesMut};
+
+use super::extension::{
+    Extension, PreSharedKeyExtension, PskKeyExchangeExtension, SupportedVersionsExtension,
+};
+use super::key_schedule::KeySchedule;
+use super::messages::{ClientHello, Finished, HandshakeMessage, HandshakeType};
+use super::record_layer::{RecordLayer, RecordType};
+use super::utils::{HASH_LENGTH, PRE_SHARED_KEY_ID, PSK_MODE_KE, VERSION_TLS_1_3};
+use anyhow::Result;
+use rc_crypto::{digest::Algorithm, hmac::SigningKey, rand};
+//
+// Client side State-machine for TLS Handshake Management.
+//
+// Internally, we manage the TLS connection by explicitly modelling the
+// client and server state-machines from RFC8446.  You can think of
+// these `State` objects as little plugins for the `Connection` class
+// that provide different behaviours of `send` and `receive` depending
+// on the state of the connection.
+//
+pub enum ClientState<F: FnMut(&[u8]) -> Result<()>> {
+    Invalid,
+    ClientWaitEe(ClientWaitEe<F>),
+    ClientWaitSh(ClientWaitSh<F>),
+    ClientWaitFinished(ClientWaitFinished<F>),
+    ClientConnected(ClientConnected<F>),
+}
+
+impl<F: FnMut(&[u8]) -> Result<()>> ClientState<F> {
```

Comment:
You've got a lot of generics going on here with `F`, and IIUC it's entirely so you can propagate a callback down to the `RecordLayer`. Without thinking too hard about it, I wonder if you could simplify the code here by:

* Have `RecordLayer` hold a `Box<dyn FnMut(&[u8]) -> Result<()>` rather than being generic over the callback (if you haven't done any boxing of callbacks in rust before, [this stackoverflow answer](https://stackoverflow.com/questions/41081240/idiomatic-callbacks-in-rust) has a decent summary under "function trait objects" header).
* Have the `ClientState` accept an already-created `RecordLayer` instance rather than accepting a callback and creating a `RecordLayer` out of it.

At the top-level of your code, you would build up the connection in a sequence of layers: make the raw I/O callbacks, then wrap them in a record layer, then wrap that in the connection state-machine, then use it for communication.

(If you try that and hate it, or have already tried it, then no big deal - just a suggestion based on the fact that all the `<F>` in here made it a bit harder to understand what was going on).

Reply:
If you prefer to keep the generics here, you could also consider defining a custom trait like `SendCallback` or similar. Giving it a meaningful name could significantly improve the readability of the code.

Reply:
Wow that.... I swear I shoulda thought of that 😂 I spent a non-trivial amount of time **typing** the trait. Thanks Ryan!!!

Reply:
Heh, I find rust is full of "attractive nuisances" like this, where there's a very clever and zero-overhead abstraction that you can come up with if you just get all the types lined up correctly...or you could put it in a box and pay a trivial amount of runtime overhead and move on with your life. I find I have to periodically ask myself "could I just clone this?" and "could I just put this in a box?" or I end up off in the weeds 😆.

Reply:
Haha I can relate to that. Like when I was starting this off, I had this crazy idea that it's technically possible to do the whole message parsing/writing without a single clone or heap allocations (or maybe just one fixed-length allocation).... That idea didn't last long (Although maybe in the future someone can attempt it 😃, pretty sure I watched a talk about someone doing something like that)